### ISSUE TITLE
cryocare prediction - output all the same tomo

### ISSUE THREAD

**Issue by frozenfas** [scipion-em/scipion-em-cryocare#30]
Hello, I have an issue with the cryocare prediction protocol.  

My pipeline so far is that I use motioncorr (align movies, even/odd =yes) ->make a subset of 3 TS -> aretomo (aln tilt series) -> imod reconstruct (even/odd =yes) -> cryocare training (are even/odd associated to tomo = no; tomograms = output from imod reconstruct).

First, I am a little confused by the question in cryocare training "are even/odd associated to tomo" as I think if I use imod, I generate tomos with the corresponding even/odd TOMO associated. In any case, if I answer no, I have the possibility to associate a single TOMO object ( with even/odd), which is what I want. Also, in the generated JSON file (training_data_config), I can see that it is correctly finding the right even/odd tomos. So this is not a problem. I just find the wording of the question in the GUI strange. Note there is also no help text to understand the question.

The problem for me is when I run the cryocare protedict protocol (on a subset of 3 TOMO and the same used for training) all three outputs are the same TOMO. I mean, in the extra directory, it correctly creates 3 directories with 3 unique names as expected, but in each of the 3 directories, the actual TOMO file is the same one:

<img width="1323" alt="Screenshot 2023-12-27 at 07 41 22" src="https://github.com/scipion-em/scipion-em-cryocare/assets/15699753/59d177ac-34f1-4cb7-a46d-35ba1172301f">

In the JSON file created for each tomo, I can also see that it is always using the even/odd for the same tomo (TEM_20210128_SSTL_FIB_20210126_SSTL_Goslar_pos3_TL_L2T2).  Visually inspecting the tomo also confirms they are the same and it is not just a naming issue:
<img width="1364" alt="Screenshot 2023-12-27 at 07 52 47" src="https://github.com/scipion-em/scipion-em-cryocare/assets/15699753/b1200e2c-63e2-4183-8b26-08b63e609f11">

The Metadate viewer does have unique TsId though for each of the 3
<img width="747" alt="Screenshot 2023-12-27 at 07 54 00" src="https://github.com/scipion-em/scipion-em-cryocare/assets/15699753/f1f449b2-aab3-40ce-9195-50c86b59b142">

Am I doing something wrong in my pipeline that is confusing this protocol? For example, is the problem that I misunderstand the first question, "are even/odd associated to tomo" and I should be answering "yes"

In case it is important, the data I am using is from EMPIAR 10860.

I am using the devel version of the plugin:

`00008:   plugin: cryocare
00009:   plugin v: 3.2.1 (devel)
00010:   plugin binary v:`

Thanks in advance for your advice.

**Comment by frozenfas** [comment]
This does not solve the problem but in protocol_predict.py

In line 149 "self.preparePredictStep(tsId, odd, even)" I think it is passing the even/odd variables in the reverse order to the "preparePredictStep(self, tsId, evenTomo, oddTomo)" function (line 151) 

Namely, maybe line 149 should be 
self.preparePredictStep(tsId, even, odd)

**Comment by frozenfas** [comment]
I still dont understand why the code is not making output as I expect but assuming I am not making a mistake in my pipeline somewhere if I change lines 131-136:

 ```
       else:
            for t in self.tomo.get():
                tsId = t.getTsId()
                # modification next three lines by sean
                fnodd, fneven = t.getHalfMaps().split(',')
                self._insertFunctionStep(self.preparePredictStep, tsId, fneven, fnodd)
                #self._insertFunctionStep(self.preparePredictFullTomoStep, tsId, t)
                self._insertFunctionStep(self.predictStep, tsId)
                self._insertFunctionStep(self.createOutputStep, tsId)
```
it generates config json files that are as I expect.  

Thanks in advance for your help and sorry if I just made a mess in my pipeline somewhere :)

**Comment by JorMaister** [comment]
Hi, @frozenfas,

First of all, many thanks for reporting. We also noticed that and there's a branch fixing it, but it has not been merged yet. The behaviuor observed is related to some memory re-usage from the core. If you are curious, these are the changes proposed in the pull request I made that should fix that.

https://github.com/scipion-em/scipion-em-cryocare/pull/29/files

On the other hand, regarding the "are even/odd associated to tomo"  question that appears in both training and prediction protocols, it's used to determine where to read the even/odd tomograms from. They can be an attribute in the tomograms set (in which each tomogram element would contain the path to the corresponding even/odd versions) or be directly introduced. Any suggestions to make it clearer for the users to be interpreted correctly are more than welcome.
